// wxt@0.20.27/dist/core/utils/create-file-reloader.mjs
import { getEntrypointBundlePath, isHtmlEntrypoint } from "./entrypoints.mjs";
import { wxt } from "../wxt.mjs";
import { detectDevChanges } from "./building/detect-dev-changes.mjs";
import { findEntrypoints } from "./building/find-entrypoints.mjs";
import { getContentScriptJs, mapWxtOptionsToRegisteredContentScript } from "./content-scripts.mjs";
import { getContentScriptCssFiles, getContentScriptsCssMap } from "./manifest.mjs";
import { rebuild } from "./building/rebuild.mjs";
import "./building/index.mjs";
import { isBabelSyntaxError, logBabelSyntaxError } from "./syntax-errors.mjs";
import { relative } from "node:path";
import { styleText } from "node:util";
import { debounce } from "perfect-debounce";
import { Mutex } from "async-mutex";
//#region src/core/utils/create-file-reloader.ts
/**
* Returns a function responsible for reloading different parts of the extension
* when a file changes.
*/
function createFileReloader(server) {
	const fileChangedMutex = new Mutex();
	const changeQueue = [];
	const cb = async (event, path) => {
		changeQueue.push([event, path]);
		await fileChangedMutex.runExclusive(async () => {
			if (server.currentOutput == null) return;
			const fileChanges = changeQueue.splice(0, changeQueue.length).map(([_, file]) => file);
			if (fileChanges.length === 0) return;
			await wxt.reloadConfig();
			const changes = detectDevChanges(fileChanges, server.currentOutput);
			if (changes.type === "no-change") return;
			if (changes.type === "full-restart") {
				wxt.logger.info("Config changed, restarting server...");
				server.restart();
				return;
			}
			if (changes.type === "browser-restart") {
				wxt.logger.info("Runner config changed, restarting browser...");
				server.restartBrowser();
				return;
			}
			wxt.logger.info(`Changed: ${Array.from(new Set(fileChanges)).map((file) => styleText("dim", relative(wxt.config.root, file))).join(", ")}`);
			const allEntrypoints = await findEntrypoints();
			try {
				const { output: newOutput } = await rebuild(allEntrypoints, changes.rebuildGroups, changes.cachedOutput);
				server.currentOutput = newOutput;
				switch (changes.type) {
					case "extension-reload":
						server.reloadExtension();
						wxt.logger.success(`Reloaded extension`);
						break;
					case "html-reload":
						const { reloadedNames } = reloadHtmlPages(changes.rebuildGroups, server);
						wxt.logger.success(`Reloaded: ${getFilenameList(reloadedNames)}`);
						break;
					case "content-script-reload":
						reloadContentScripts(changes.changedSteps, server);
						const rebuiltNames = changes.rebuildGroups.flat().map((entry) => entry.name);
						wxt.logger.success(`Reloaded: ${getFilenameList(rebuiltNames)}`);
						break;
				}
			} catch {}
		}).catch((error) => {
			if (!isBabelSyntaxError(error)) throw error;
			logBabelSyntaxError(error);
		});
	};
	return debounce(cb, wxt.config.dev.server.watchDebounce, {
		leading: true,
		trailing: false
	});
}
/**
* From the server, tell the client to reload content scripts from the provided
* build step outputs.
*/
function reloadContentScripts(steps, server) {
	if (wxt.config.manifestVersion === 3) steps.forEach((step) => {
		if (server.currentOutput == null) return;
		const entry = step.entrypoints;
		if (Array.isArray(entry) || entry.type !== "content-script") return;
		const js = getContentScriptJs(wxt.config, entry);
		const cssMap = getContentScriptsCssMap(server.currentOutput, [entry]);
		const css = getContentScriptCssFiles([entry], cssMap);
		server.reloadContentScript({
			registration: entry.options.registration,
			contentScript: mapWxtOptionsToRegisteredContentScript(entry.options, js, css)
		});
	});
	else server.reloadExtension();
}
function reloadHtmlPages(groups, server) {
	const htmlEntries = groups.flat().filter(isHtmlEntrypoint);
	htmlEntries.forEach((entry) => {
		const path = getEntrypointBundlePath(entry, wxt.config.outDir, ".html");
		server.reloadPage(path);
	});
	return { reloadedNames: htmlEntries.map((entry) => entry.name) };
}
function getFilenameList(names) {
	return names.map((name) => styleText("cyan", name)).join(styleText("dim", ", "));
}
//#endregion
export { createFileReloader, reloadContentScripts };
